### ISSUE TITLE
Arrows'/projectile spells' effective chance to hit influenced by projectile speed

### ISSUE THREAD

**Issue by SoundChaser83** [diasurgical/devilutionX#2826]
**Important information**
Operating system / DevilutionX version / did you download a release or did you compile yourself?

https://github.com/diasurgical/devilutionX/commit/effee3f7bb7b3e3b868befb2006d3200e5e617a5

**Describe**
Slower missiles often perform a hit check multiple times as they spend time passing over a monster's tile, meaning that a projectile that should have low chance to hit gets multiple attempts to roll a hit and therefore has higher _effective_ chance to hit. This seems to be the more common case. However, faster missiles can sometimes skip a hit check altogether seemingly by never spending time on the same tile as a monster, meaning that faster missiles have lower _effective_ chance to hit, other things equal.

This can be tested by tracking the `hper` variable within the `MonsterMHit` method in missiles.cpp while running the game and shooting stuff with arrows and things like Firebolt. You can exaggerate the problem by altering the speed of arrows/firebolts to be extremely slow and by drastically reducing your (normal) hit chance, and more hit checks will be output even when firing single shots at one monster.

This seems like it might be a vanilla bug as well, but I'm unsure. Also possibly related to #1672 and #1531

**To Reproduce**
1. Get a character with bad chance to hit for magic or bows.
2. Shoot monsters.
3. Compare actual chance to successfully hit a monster with expected chance to hit.

**Expected behavior**
A projectile performs exactly one hit check against a monster it collides with. If it fails the first hit check, it continues past the initial monster and cannot attempt to hit it again. Faster projectiles should also always perform a hit check against a monster standing in the projectile's path.

**Screenshots**
![sorctesting](https://user-images.githubusercontent.com/54762359/132428037-cfff4892-17dd-44b0-bdd3-4bd18f0048dd.png)
![hitchancedebug](https://user-images.githubusercontent.com/54762359/132428094-a6ccb88b-774e-4c05-a8dc-57e22c7d8eb4.png)
I drastically lowered Sorcerer's hit chance in the code and shot some stuff on dlvl 1. The debug screenshot shows my chance to hit being returned on each hit check against a monster I shot at, and it also shows me tapping spacebar in between each shot. On the final shot against the Scavenger a couple tiles directly south of me, the game performed three hit checks as seen by the three 46s in a row, and ultimately landing the hit and killing it on the third hit check. This is with a normal speed Firebolt spell.

**Comment by julealgon** [comment]
Reading this now i actually wonder how we'd fix the "high speed no hit" problem.... since the game only ticks at 20 fps internally, i guess we'd have to do some predictive collision detection in that case. It almost seems like we are dealing with 2 distinct issues here if I'm honest 

I guess it would depend on how exactly we end up fixing this.

**Comment by julealgon** [comment]
> This is essentially a duplicate of #1531. At least, the root cause is the same between these two issues.

This better describes the problem at its core than the blocking issue, which is basically a consequence of this and/or might be slightly different since this talks about "multiple hit checks" while that is about a missile that is not being destroyed after being blocked.

Sure we can fix both with the same fix, but i think closing this one is a mistake. Either keep both open, or close the other one instead.

My personal vote would be to keep both open since they are actually not 100% equivalent.

**Comment by StephenCWills** [comment]
> Either keep both open, or close the other one instead.

As long as one references the other, I don't really care which stays open. I just gave priority to the first one that was created.

**Comment by qndel** [comment]
Actually like this one because it describes the problem well. If we make missiles hit something only once, we have to remember to allow multihit in certain cases (firewall / lightning wall / lightning etc.)

**Comment by NiteKat** [comment]
Do missiles store their previous x and y? I don't think they do... if they did, you could add a check for if either x or y has changed before doing a check on monster hits? With exceptions for stationary spells like firewall / lightning wall, and I guess lightning should be able to hit multiple times too?

Oh hey, could this be related to elemental arrows doing crap loads of damage sometimes too?

**Comment by StephenCWills** [comment]
> Do missiles store their previous x and y? I don't think they do... if they did, you could add a check for if either x or y has changed before doing a check on monster hits? With exceptions for stationary spells like firewall / lightning wall, and I guess lightning should be able to hit multiple times too?
>
> ...
>
> Edit: I guess this wouldn't be fool proof. If you looked for x and y of missile to change, then you could safely step on a tile that is currently occupied by a projectile, changing gameplay quite a bit...

There is one thing this issue highlights that the other issue does not. Eliminating the multiple hit chance of moving projectiles could significantly reduce their accuracy. That alone could significantly affect the game balance.

> Oh hey, could this be related to elemental arrows doing crap loads of damage sometimes too?

Not exactly. That does occur because the missile collides multiple times, but it's because the missile isn't deleted after the first successful collision. See #1672.

**Comment by StephenCWills** [comment]
Just a few more notes that are potentially worth mentioning.

Apart from projectile speed, the path that a projectile takes through any given tile will also help determine how many hit checks the projectile has in that tile. So even if your target is not moving, they may be more difficult to hit from certain angles than others. Furthermore, enemies at awkward angles could be more difficult to hit when aiming projectiles at different tiles behind them.

Even apart from speed and angle, distance to the target could also affect the number of hit chances. To provide a simple example, if a projectile is moving at 0.9 tiles per tick, every tenth tile it passes through would get two hit checks instead of one. Conversely, if a projectile is moving at 1.1 tiles per tick, every tenth tile will get no hit checks! 😮 

Lastly, the tickrate of the game has a big impact on this as well. If someone were to decide to increase the game's tickrate to 60 fps for their mod, it would inadvertently increase the number of hit chances for moving projectiles threefold.

What a mess.

**Comment by AJenbo** [comment]
How about only allowing one impact per tile?
```
if (missile.pointOfLastImpact != missile.position.tile) {
  if (CheckMissileCol())
    missile.pointOfLastImpact = missile.position.tile;
}
```

**Comment by StephenCWills** [comment]
> How about only allowing one impact per tile?

I think that's a good first step. Ideally, we would also interpolate the projectile's path between ticks when handling collisions and adjust the accuracy of projectiles to compensate for the loss of accuracy. At the very least, I feel like the current situation is too insane to handle it any other way.

**Comment by AJenbo** [comment]
Now we just need to also fix firewalls not being continious when placed at an angle :D

**Comment by julealgon** [comment]
> What a mess.

I like this TL;DR.

This might be one of the nastiest bugs in the game if you think about it...

**Comment by NiteKat** [comment]
Another thing to note with projectiles - certain tiles can be missed altogether. I've noticed this with the AI I have that plays the game - if an enemy is far enough away, but one of the axis is only a 1 tile difference, arrows will sail right past... and it isn't because there's no chance to hit. The missile either is missing that tile, or maybe due to tick rate doesn't have a hit check on that tile? It's quite annoying when it happens! Human players don't tend to fire from that far away, so you don't usually see this in human play.

**Comment by julealgon** [comment]
> Human players don't tend to fire from that far away, so you don't usually see this in human play.

This will change dramatically with the support for variable internal resolution. Most people prefer playing with a larger play area.

Would be nice if you could share specific details about these situations so we could repro them.

**Comment by NiteKat** [comment]
> Would be nice if you could share specific details about these situations so we could repro them.
Ya know... I thought I had opened an issue for or mentioned this somewhere, but I don't see it in https://github.com/diasurgical/devilution/issues/64 or anywhere on the DevX repo, so next time the AI stumbles on the situation, I'll make a save file and add it somewhere and reference it here.

**Comment by NiteKat** [comment]
> Would be nice if you could share specific details about these situations so we could repro them.

Opened up an issue for it. https://github.com/diasurgical/devilutionX/issues/2837 :)
